File: src/domain/rules.ts
// The four business rules, as pure functions: no I/O, `today` always passed in
// explicitly, results are structured detail rather than booleans. Kept free of
// React/Dexie/network deps (enforced by eslint) so both the mock API and the
// forms can import the same logic.

import { daysBetween } from './dates';
import {
  type Assignment,
  type Certification,
  type CertificationType,
  type CrewMember,
  type IsoDate,
  RANKS,
  type Rank,
  type Vessel,
} from './types';

/** Certificates required per rank for an assignment to be legal (rule 3). Policy, not spec-derived. */
export const REQUIRED_CERTIFICATIONS_BY_RANK: Record<Rank, readonly CertificationType[]> = {
  Master: ['STCW', 'Medical Fitness', 'GMDSS', "Seaman's Book", 'Passport'],
  'Chief Officer': ['STCW', 'Medical Fitness', 'GMDSS', "Seaman's Book", 'Passport'],
  'Chief Engineer': ['STCW', 'Medical Fitness', "Seaman's Book", 'Passport'],
  '2nd Engineer': ['STCW', 'Medical Fitness', "Seaman's Book", 'Passport'],
  AB: ['STCW', 'Medical Fitness', "Seaman's Book"],
  Oiler: ['STCW', 'Medical Fitness', "Seaman's Book"],
  Cook: ['Medical Fitness', "Seaman's Book"],
};

/** Half-open `[start, end)`: a same-day handover (one ends where the next begins) is legal. */
function rangesOverlap(aStart: IsoDate, aEnd: IsoDate, bStart: IsoDate, bEnd: IsoDate): boolean {
  return !(aEnd <= bStart || bEnd <= aStart);
}

/** Existing assignments that conflict with `candidate`: same crew, Planned/Active, not itself, overlapping dates. */
export function findConflictingAssignments(
  candidate: Assignment,
  existing: readonly Assignment[],
): Assignment[] {
  return existing.filter((assignment) => {
    const sameCrewMember = candidate.crewId === assignment.crewId;
    const isNotCandidateItself = candidate.id !== assignment.id;
    const occupiesCalendar = assignment.status === 'Active' || assignment.status === 'Planned';
    const datesConflict = rangesOverlap(
      candidate.signOnDate,
      candidate.signOffDate,
      assignment.signOnDate,
      assignment.signOffDate,
    );

    return sameCrewMember && isNotCandidateItself && occupiesCalendar && datesConflict;
  });
}

export interface ManningShortfall {
  rank: Rank;
  required: number;
  actual: number;
  short: number;
}

export interface ManningCompliance {
  compliant: boolean;
  shortfalls: ManningShortfall[];
}

/** Compares a vessel's currently-crewed ranks (Active, today within range) against its minimum manning. */
export function checkManningCompliance(
  vessel: Vessel,
  assignments: readonly Assignment[],
  today: IsoDate,
): ManningCompliance {
  const activeRoster = assignments.filter((assignment) => {
    const validVesselId = assignment.vesselId === vessel.id;
    const validStatus = assignment.status === 'Active';
    const todayValid = assignment.signOnDate <= today && today <= assignment.signOffDate;
    return validVesselId && todayValid && validStatus;
  });

  const countByRank: Partial<Record<Rank, number>> = {};
  for (const assignment of activeRoster) {
    countByRank[assignment.rankOnboard] = (countByRank[assignment.rankOnboard] ?? 0) + 1;
  }

  const shortfalls: ManningShortfall[] = [];
  for (const rank of RANKS) {
    const required = vessel.minimumSafeManning[rank] ?? 0;
    const actual = countByRank[rank] ?? 0;
    if (actual < required) {
      shortfalls.push({ rank, required, actual, short: required - actual });
    }
  }

  return { compliant: shortfalls.length === 0, shortfalls };
}

export interface CertificationBlock {
  type: CertificationType;
  reason: 'missing' | 'expires-before-sign-off';
  expiryDate?: IsoDate;
  daysShort?: number;
}

/** Certificates blocking this rotation: missing, or expiring before `signOffDate` (not "today"). */
export function findBlockingCertifications(
  crew: CrewMember,
  certifications: readonly Certification[],
  rankOnboard: Rank,
  signOffDate: IsoDate,
): CertificationBlock[] {
  const requiredTypes = REQUIRED_CERTIFICATIONS_BY_RANK[rankOnboard];
  const crewCertifications = certifications.filter((cert) => cert.crewId === crew.id);

  const blocks: CertificationBlock[] = [];
  for (const type of requiredTypes) {
    // Multiple certs of the same type can exist (e.g. a renewal); use whichever expires latest.
    const matching = crewCertifications.filter((cert) => cert.type === type);
    const best = matching.reduce<Certification | undefined>((latest, cert) => {
      if (!latest || cert.expiryDate > latest.expiryDate) return cert;
      return latest;
    }, undefined);

    if (!best) {
      blocks.push({ type, reason: 'missing' });
      continue;
    }

    if (best.expiryDate < signOffDate) {
      blocks.push({
        type,
        reason: 'expires-before-sign-off',
        expiryDate: best.expiryDate,
        daysShort: daysBetween(best.expiryDate, signOffDate),
      });
    }
  }

  return blocks;
}

/** Days an Active assignment has overrun `signOffDate`; 0 if not overdue or not Active. */
export function getOverdueDays(assignment: Assignment, today: IsoDate): number {
  if (assignment.status !== 'Active') return 0;

  const overdueDays = daysBetween(assignment.signOffDate, today);
  return overdueDays > 0 ? overdueDays : 0;
}

File: server/CrewLink.Api/Domain/Rules.cs
using System.Collections.Frozen;
using System.Text.Json.Serialization;

namespace CrewLink.Api.Domain;

/// <summary>
/// The four business rules, as pure functions: no I/O, <c>today</c> always passed
/// in explicitly, results are structured detail rather than booleans.
/// </summary>
/// <remarks>
/// A deliberate second implementation of <c>src/domain/rules.ts</c> — the browser
/// build has to work with no backend, so the logic exists twice, guarded against
/// drift by <c>RulesTests.cs</c> porting <c>rules.test.ts</c> case for case.
/// </remarks>
public static class Rules
{
    /// <summary>
    /// Certificates required per rank for an assignment to be legal (rule 3).
    /// Policy, not spec-derived — the spec never defines "any required
    /// certification", so this fills the gap explicitly.
    /// </summary>
    public static readonly FrozenDictionary<Rank, CertificationType[]> RequiredCertificationsByRank =
        new Dictionary<Rank, CertificationType[]>
        {
            [Rank.Master] =
            [
                CertificationType.Stcw, CertificationType.MedicalFitness, CertificationType.Gmdss,
                CertificationType.SeamansBook, CertificationType.Passport,
            ],
            [Rank.ChiefOfficer] =
            [
                CertificationType.Stcw, CertificationType.MedicalFitness, CertificationType.Gmdss,
                CertificationType.SeamansBook, CertificationType.Passport,
            ],
            [Rank.ChiefEngineer] =
            [
                CertificationType.Stcw, CertificationType.MedicalFitness,
                CertificationType.SeamansBook, CertificationType.Passport,
            ],
            [Rank.SecondEngineer] =
            [
                CertificationType.Stcw, CertificationType.MedicalFitness,
                CertificationType.SeamansBook, CertificationType.Passport,
            ],
            [Rank.AB] =
            [
                CertificationType.Stcw, CertificationType.MedicalFitness, CertificationType.SeamansBook,
            ],
            [Rank.Oiler] =
            [
                CertificationType.Stcw, CertificationType.MedicalFitness, CertificationType.SeamansBook,
            ],
            [Rank.Cook] = [CertificationType.MedicalFitness, CertificationType.SeamansBook],
        }.ToFrozenDictionary();

    /// <summary>
    /// Half-open <c>[start, end)</c>: a same-day handover is legal. A closed
    /// interval is equally defensible — this is a judgment call, pinned by a test.
    /// </summary>
    private static bool RangesOverlap(DateOnly aStart, DateOnly aEnd, DateOnly bStart, DateOnly bEnd) =>
        !(aEnd <= bStart || bEnd <= aStart);

    /// <summary>
    /// Rule 1. Existing assignments that conflict with <paramref name="candidate"/>:
    /// same crew member, Planned or Active, not the candidate itself, overlapping dates.
    /// </summary>
    public static List<Assignment> FindConflictingAssignments(
        Assignment candidate,
        IEnumerable<Assignment> existing)
    {
        var conflicts = new List<Assignment>();
        foreach (var assignment in existing)
        {
            var sameCrewMember = candidate.CrewId == assignment.CrewId;
            var isNotCandidateItself = candidate.Id != assignment.Id;
            var occupiesCalendar = assignment.Status is AssignmentStatus.Active or AssignmentStatus.Planned;
            var datesConflict = RangesOverlap(
                candidate.SignOnDate, candidate.SignOffDate,
                assignment.SignOnDate, assignment.SignOffDate);

            if (sameCrewMember && isNotCandidateItself && occupiesCalendar && datesConflict)
            {
                conflicts.Add(assignment);
            }
        }
        return conflicts;
    }

    /// <summary>
    /// Rule 2. Compares a vessel's currently-crewed ranks against its minimum safe
    /// manning. "Currently crewed" is Active with <paramref name="today"/> inside
    /// the contract dates, inclusive on both ends — unlike rule 1's half-open choice.
    /// </summary>
    public static ManningCompliance CheckManningCompliance(
        Vessel vessel,
        IEnumerable<Assignment> assignments,
        DateOnly today)
    {
        // Counts the rank actually sailed, not the crew member's substantive
        // rank — a Chief Officer sailing as Master fills the Master slot.
        var countByRank = new Dictionary<Rank, int>();
        foreach (var assignment in assignments)
        {
            var validVesselId = assignment.VesselId == vessel.Id;
            var validStatus = assignment.Status == AssignmentStatus.Active;
            var todayValid = assignment.SignOnDate <= today && today <= assignment.SignOffDate;
            if (!validVesselId || !validStatus || !todayValid) continue;

            countByRank[assignment.RankOnboard] = countByRank.GetValueOrDefault(assignment.RankOnboard) + 1;
        }

        var shortfalls = new List<ManningShortfall>();
        foreach (var rank in Ranks.All)
        {
            var required = vessel.MinimumSafeManning.GetValueOrDefault(rank);
            var actual = countByRank.GetValueOrDefault(rank);
            if (actual < required)
            {
                shortfalls.Add(new ManningShortfall(rank, required, actual, required - actual));
            }
        }

        return new ManningCompliance(shortfalls.Count == 0, shortfalls);
    }

    /// <summary>
    /// Rule 3. Certificates blocking this rotation: missing outright, or expiring
    /// before <paramref name="signOffDate"/> — not before "today". Valid up to and
    /// including its expiry date, so expiring exactly on sign-off doesn't block.
    /// </summary>
    public static List<CertificationBlock> FindBlockingCertifications(
        CrewMember crew,
        IEnumerable<Certification> certifications,
        Rank rankOnboard,
        DateOnly signOffDate)
    {
        var held = certifications.Where(certification => certification.CrewId == crew.Id).ToList();
        var blocks = new List<CertificationBlock>();

        foreach (var type in RequiredCertificationsByRank[rankOnboard])
        {
            // Two certificates of the same type can exist (an old one and a
            // renewal) — take whichever expires latest.
            Certification? best = null;
            foreach (var certification in held)
            {
                if (certification.Type != type) continue;
                if (best is null || certification.ExpiryDate > best.ExpiryDate) best = certification;
            }

            if (best is null)
            {
                blocks.Add(new CertificationBlock(type, CertificationBlockReason.Missing));
                continue;
            }

            if (best.ExpiryDate < signOffDate)
            {
                blocks.Add(new CertificationBlock(
                    type,
                    CertificationBlockReason.ExpiresBeforeSignOff,
                    best.ExpiryDate,
                    Dates.DaysBetween(best.ExpiryDate, signOffDate)));
            }
        }

        return blocks;
    }

    /// <summary>
    /// Rule 4. Days an Active assignment has overrun its sign-off date; 0 when not
    /// overdue. A Planned assignment past its dates is a planning error, not an
    /// overrun, and returns 0 rather than conflating the two.
    /// </summary>
    public static int GetOverdueDays(Assignment assignment, DateOnly today)
    {
        if (assignment.Status != AssignmentStatus.Active) return 0;

        var overdueDays = Dates.DaysBetween(assignment.SignOffDate, today);
        return overdueDays > 0 ? overdueDays : 0;
    }
}

public sealed record ManningShortfall(Rank Rank, int Required, int Actual, int Short);

public sealed record ManningCompliance(bool Compliant, IReadOnlyList<ManningShortfall> Shortfalls);

public enum CertificationBlockReason
{
    [JsonStringEnumMemberName("missing")] Missing,
    [JsonStringEnumMemberName("expires-before-sign-off")] ExpiresBeforeSignOff,
}

public sealed record CertificationBlock(
    CertificationType Type,
    CertificationBlockReason Reason,
    DateOnly? ExpiryDate = null,
    int? DaysShort = null);
